File: shim/cpp/shim.cpp
#include "shim.h"

#include "libchess/uci.h"

#include <chrono>
#include <iostream>
#include <sstream>
#include <string>

namespace shim {

using namespace libchess;

namespace {

// Rebuild the board from a `position` command:
//   position startpos [moves ...]
//   position fen <6 fields> [moves ...]
void apply_position(Board& board, std::istringstream& ss) {
    std::string token;
    ss >> token;
    if (token == "startpos") {
        board.set_startpos();
    } else if (token == "fen") {
        std::string fen, part;
        for (int i = 0; i < 6 && (ss >> part); ++i) {
            fen += part;
            if (i < 5) fen += ' ';
        }
        board.set_fen(fen);
    }
    // Optional moves.
    while (ss >> token) {
        if (token == "moves") continue;
        Move m = from_uci(board, token);
        if (m != MOVE_NONE) board.make_move(m);
    }
}

Limits parse_go(std::istringstream& ss) {
    Limits lim;
    std::string token;
    while (ss >> token) {
        if (token == "movetime") ss >> lim.ms_left;
        else if (token == "nodes") ss >> lim.nodes_left;
        // Other UCI go args (wtime/btime/depth/infinite) are ignored in Phase 0.
    }
    return lim;
}

} // namespace

int run() {
    std::ios::sync_with_stdio(false);
    Board board;
    Meta m = meta();

    std::string line;
    while (std::getline(std::cin, line)) {
        std::istringstream ss(line);
        std::string cmd;
        ss >> cmd;

        if (cmd == "uci") {
            std::cout << "id name " << m.name << "\n"
                      << "id author chess-world-cup\n"
                      << "id lang " << m.lang << "\n"
                      << "id family " << m.family << "\n"
                      << "id country " << m.country << "\n"
                      << "uciok\n";
            std::cout.flush();
        } else if (cmd == "isready") {
            std::cout << "readyok\n";
            std::cout.flush();
        } else if (cmd == "ucinewgame") {
            board.set_startpos();
        } else if (cmd == "position") {
            apply_position(board, ss);
        } else if (cmd == "go") {
            Limits lim = parse_go(ss);
            uint64_t nodes = 0;
            auto t0 = std::chrono::steady_clock::now();
            Move best = search(board, lim, nodes);
            auto t1 = std::chrono::steady_clock::now();
            long long ms =
                std::chrono::duration_cast<std::chrono::milliseconds>(t1 - t0).count();
            // Self-reported search time & nodes — the orchestrator subtracts this
            // from its own measurement to isolate the implementation tax.
            std::cout << "info time " << ms << " nodes " << nodes << "\n"
                      << "bestmove " << to_uci(best) << "\n";
            std::cout.flush();
        } else if (cmd == "quit") {
            break;
        }
    }
    return 0;
}

} // namespace shim

File: shim/py/shim.py
"""Python UCI shim — mirrors shim/cpp. A bot provides meta() and search(); the
shim handles the UCI dance, position/FEN parsing, time bookkeeping and the
self-reported timing line. Ship-a-new-language contract: only those two funcs.
"""
import sys
import time

from libchess_ffi import Board


class Limits:
    __slots__ = ("ms_left", "nodes_left")

    def __init__(self, ms_left=-1, nodes_left=-1):
        self.ms_left = ms_left
        self.nodes_left = nodes_left


def _apply_position(board, tokens):
    i = 0
    if tokens and tokens[0] == "startpos":
        board.startpos()
        i = 1
    elif tokens and tokens[0] == "fen":
        fen = " ".join(tokens[1:7])
        board.set_fen(fen)
        i = 7
    if i < len(tokens) and tokens[i] == "moves":
        for uci in tokens[i + 1:]:
            m = board.move_from_uci(uci)
            if m:
                board.make(m)


def _parse_go(tokens):
    lim = Limits()
    i = 0
    while i < len(tokens):
        if tokens[i] == "movetime" and i + 1 < len(tokens):
            lim.ms_left = int(tokens[i + 1]); i += 2
        elif tokens[i] == "nodes" and i + 1 < len(tokens):
            lim.nodes_left = int(tokens[i + 1]); i += 2
        else:
            i += 1
    return lim


def run(meta, search):
    board = Board()
    m = meta()
    out = sys.stdout
    for line in sys.stdin:
        parts = line.split()
        if not parts:
            continue
        cmd = parts[0]
        if cmd == "uci":
            out.write(f"id name {m['name']}\n")
            out.write("id author chess-world-cup\n")
            out.write(f"id lang {m['lang']}\n")
            out.write(f"id family {m['family']}\n")
            out.write(f"id country {m['country']}\n")
            out.write("uciok\n"); out.flush()
        elif cmd == "isready":
            out.write("readyok\n"); out.flush()
        elif cmd == "ucinewgame":
            board.startpos()
        elif cmd == "position":
            _apply_position(board, parts[1:])
        elif cmd == "go":
            lim = _parse_go(parts[1:])
            t0 = time.monotonic()
            best, nodes = search(board, lim)
            ms = int((time.monotonic() - t0) * 1000)
            uci = board.move_to_uci(best) if best else "0000"
            out.write(f"info time {ms} nodes {nodes}\n")
            out.write(f"bestmove {uci}\n"); out.flush()
        elif cmd == "quit":
            break
    return 0
